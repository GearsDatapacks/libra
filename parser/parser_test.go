package parser_test

import (
	"testing"

	"github.com/gearsdatapacks/libra/lexer"
	"github.com/gearsdatapacks/libra/lexer/token"
	"github.com/gearsdatapacks/libra/parser"
	"github.com/gearsdatapacks/libra/parser/ast"
	utils "github.com/gearsdatapacks/libra/test_utils"
)

func getProgram(t *testing.T, input string) *ast.Program {
	l := lexer.New(input, "test.lb")
	tokens := l.Tokenise()

	p := parser.New(tokens, l.Diagnostics)
	program := p.Parse()
	utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 0)

	return program
}

func getExpr[T ast.Expression](t *testing.T, program *ast.Program) T {
	utils.AssertEq(t, len(program.Statements), 1)
	stmt, ok := program.Statements[0].(*ast.ExpressionStatement)
	utils.Assert(t, ok)

	expr, ok := stmt.Expression.(T)
	utils.Assert(t, ok)
  return expr
}

func TestIdentifierExpression(t *testing.T) {
	input := "hello_World123"
  program := getProgram(t, input)

  ident := getExpr[*ast.Identifier](t, program)

	utils.AssertEq(t, ident.Name, input)
	utils.AssertEq(t, ident.Token,
		token.New(token.IDENTIFIER, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestIntegerExpression(t *testing.T) {
	input := "156_098"
	val := 156098

  program := getProgram(t, input)

  integer := getExpr[*ast.IntegerLiteral](t, program)

	utils.AssertEq(t, integer.Value, int64(val))
	utils.AssertEq(t, integer.Token,
		token.New(token.INTEGER, "156098",
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestFloatExpression(t *testing.T) {
	input := "3.1415"
	val := 3.1415

  program := getProgram(t, input)

  float := getExpr[*ast.FloatLiteral](t, program)

	utils.AssertEq(t, float.Value, val)
	utils.AssertEq(t, float.Token,
		token.New(token.FLOAT, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestBooleanExpression(t *testing.T) {
	input := "true"

  program := getProgram(t, input)

  boolean := getExpr[*ast.BooleanLiteral](t, program)

	utils.AssertEq(t, boolean.Value, true)
	utils.AssertEq(t, boolean.Token,
		token.New(token.IDENTIFIER, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestErrorExpression(t *testing.T) {
  input := ")"

  l := lexer.New(input, "test.lb")
  tokens := l.Tokenise()
  p := parser.New(tokens, l.Diagnostics)
  program := p.Parse()

  utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 1)
  diag := p.Diagnostics.Diagnostics[0]
  utils.AssertEq(t, diag.Message, "Expected expression, got `)`")
  utils.AssertEq(t, diag.Span, token.NewSpan(0, 0, 1))

  getExpr[*ast.ErrorExpression](t, program)
}

func TestMissingNewlineError(t *testing.T) {
  input := "1 2"

  l := lexer.New(input, "test.lb")
  tokens := l.Tokenise()
  p := parser.New(tokens, l.Diagnostics)
  p.Parse()

  utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 1)
  diag := p.Diagnostics.Diagnostics[0]
  utils.AssertEq(t, diag.Message, "Expected newline after statement, got integer")
  utils.AssertEq(t, diag.Span, token.NewSpan(0, 2, 3))
}

