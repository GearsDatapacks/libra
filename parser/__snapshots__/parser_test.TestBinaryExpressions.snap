
[`1 + 2` - 1]
BIN_EXPR + (0:5)
├─INT_LIT 1 (0:1)
└─INT_LIT 2 (4:5)
---

[`"Hello" + "world"` - 1]
BIN_EXPR + (0:17)
├─STRING_LIT "Hello" (0:7)
└─STRING_LIT "world" (10:17)
---

[`foo - bar` - 1]
BIN_EXPR - (0:9)
├─IDENT foo (0:3)
└─IDENT bar (6:9)
---

[`19 / 27` - 1]
BIN_EXPR / (0:7)
├─INT_LIT 19 (0:2)
└─INT_LIT 27 (5:7)
---

[`1 << 2` - 1]
BIN_EXPR << (0:6)
├─INT_LIT 1 (0:1)
└─INT_LIT 2 (5:6)
---

[`7 &19` - 1]
BIN_EXPR & (0:5)
├─INT_LIT 7 (0:1)
└─INT_LIT 19 (3:5)
---

[`15.04* 1_2_3` - 1]
BIN_EXPR * (0:12)
├─FLOAT_LIT 15.04 (0:5)
└─INT_LIT 123 1_2_3 (7:12)
---

[`true||false` - 1]
BIN_EXPR || (0:11)
├─BOOL_LIT true (0:4)
└─BOOL_LIT false (6:11)
---

[`[1,2,3]<< 4` - 1]
BIN_EXPR << (0:11)
├─LIST_EXPR (0:1)
│ ├─INT_LIT 1 (1:2)
│ ├─INT_LIT 2 (3:4)
│ └─INT_LIT 3 (5:6)
└─INT_LIT 4 (10:11)
---

[`21 ^ 35` - 1]
BIN_EXPR ^ (0:7)
├─INT_LIT 21 (0:2)
└─INT_LIT 35 (5:7)
---
